lib/flight.rb

```diff
@@ -31,6 +31,13 @@
 
 module Flight
   class << self
+    attr_reader :config
+
+    def load_configuration
+      @config = FlightJob::Configuration.load
+      FlightJob::Configuration.log_warnings
```

Comment:
The integration between `FlightConfiguration` and  the `Flight` stub is somewhat awkward.

`FlightConfiguration` needs `Flight.logger` setup before calling `log_warnings`. This creates some circular logic as `Flight.logger` needs `Flight.config` (thus `FlightConifguration.load`) to set itself up.

By running the two methods independently, it breaks the circular dependency.

---

The crux of the issue is `FlightConfiguration::DSL.log_warnings` has a dependency on `Flight`. Possible a better way to handle this is formally integrate logging within `FlightConfiguration`.

Possible something like this:

```
class Flight
  delegate :logger, to: :config
  
  ...
end

class Configuration
  extend FlightConfiguration:DSL
  extend FlightConfiguration::Logging
  
  ... as per normal sans logging attributes
end

class FlightConfiguration::Logging
  def self.load
    super.tap do |c|
      c.logger ...
    end
  end
  
  extended do
    attribute :log_level ...
    ...
  end
  
  def logger
     @logger ||= Logger.new(level: log_level) ...
  end
end
```

Reply:
Aha! The circular dependency you're trying to avoid is `Flight.logger` -> `Flight.config` -> `FlightConfiguration.load` -> `Flight.logger`.

> `FlightConfiguration` needs `Flight.logger` setup before calling `log_warnings`

That's because it logs the warnings.  It could instead return a string (or array of strings?) that the application can chose to log.  Then we could have

```ruby
def load_configuration
  @config = FlightJob::Configuration.load
  Flight.logger.warn(@config.warnings) if @config.has_warnings?
end
```

That looks much cleaner.

Reply:
Yep that is most certainly cleaner, but with one catch: _it defines instance methods_

So far `FlightConfiguration::DSL` has been class method heavy so we can avoid defining instance methods. This allows the underlying application to use a wider range of `attributes`. This is why I named `__sources__` using python-esque syntax. Obviously there is the inbuilt ruby `instance_methods`, so a few more would be _okay_.

---

The other problem is the possibility of expansion. It would be nice if `FlightConfiguration` made:
* `info` logs of the loaded configs, and
* `debug` logs of the `__sources__` structs.

It would be ideal if `FlightConfiguration` had access to the `logger` directly.

PS: `log_warnings` is a bad name, I only thought of this after implementing this PR.

---

Finally, most of our applications use a fairly similar logging configuration. It would nice if this was extracted into a common location. `FlightConfiguration` looks to be a suitable place for it, as it does configure the `logger`.

An opt-in mechanism is preferable as it makes retro-fitting easier. It would mean non-port applications wouldn't get the full logging, but this is already the case.

Reply:
Minimising the number of instance methods added by `FC::DSL` is a good idea.  Trying to eliminate all instances methods is likely taking that too far.

Having `FlightConfiguration` configure the logger directly might be a good idea, but its outside the scope of this PR.

Reply:
> That's because it logs the warnings. It could instead return a string (or array of strings?) that the application can chose to log.

So I started of implementing the instance methods as an array of strings as suggested. But quickly realised a  hash would be more versatile.

So we now support `info`/`debug` level logs without directly integrating with `Flight.logger`. These logs are stored within a `__logs__` object on the `config`.